### ISSUE TITLE
"Native" emscripten build environment

### ISSUE THREAD

**Issue by SoniEx2** [emscripten-core/emscripten#14453]
Emscripten should figure out a way to provide an execution environment with the following features:

- posix_spawn/system, using filesystem emulation and unsafe-eval if needed (to load wasm modules from the emulated filesystem).

This would enable building and running thngs like make and uname *in* emscripten, which in turn plays nicer with build systems, which in turn makes it possible to build more stuff. And because it's not seen as cross-compilation, the build system can avoid making assumptions such as whether fork works, and it can also generate artifacts as needed (e.g. libgcrypt/libgpg-error need to generate some stuff on the host machine, which isn't really possible with cross-compilation).

**Comment by sbc100** [comment]
Support for multiple "processes" is indeed something that emscripten lack any support for at all.     I agree it would be quite an interesting and ambitious project to add such support. 

Getting this working to extent where once can actually run "make" inside emscripten would be a lot of work indeed.   There are some folks who have tried this with some success: https://browsix.org/

**Comment by Antony74** [comment]
Here's the current state of GNU Make build by emscripten and run with nodejs is, to the best of my knowledge:

I've asked around to if anything like this already existed: [Adding make to the npm toolchain?](https://www.reddit.com/r/node/comments/1fmac0o/adding_make_to_the_npm_toolchain/)

And no-one came up with anything.  This also gave me the opportunity to defend this only-slightly-crazy idea on two different fronts.

GNU Make builds fine with emscripten, but running it with a very simple Makefile with nodejs on Windows we get

    make.js: getcwd: No such file or directory
    node reverse.js result.json input.json
    make.js: *** [Makefile:2: all] Error 127

The getcwd thing is a curious side-effect of using what the build system thinks is a Posix build on a Windows file system and is [easily patched on the JavaScript side](https://github.com/Antony74/make.wasm/commit/7e88f31c314a2a2024cde3f831aad88b1cbf79e0).

The error is caused by the fact posix_spawn is stubbed as follows

    weak int posix_spawn(pid_t *pid, const char *path,
                       const posix_spawn_file_actions_t *file_actions,
                       const posix_spawnattr_t *attrp,
                       char *const argv[], char *const envp[]) {
      errno = ENOSYS;
      return -1;
    }

My main worry is what else might be lurking in the GNU Make source code which rather than giving a compile-time error, is stubbed by Emscripten to give a run-time error.  We might not get to find out until a user falls over an obsure piece of Make functionality.

It looks possible to me to convert the way this is stubbed to the way getcwd is stubbed, and implement it on the nodejs side with `child_process.spawn` - I mean the difficultly is above average because there's callbacks for stdout, stderror, and an exit code, but I think it should probably be within my abilities.  Currently I've only got as far as building LLVM as it's literal hours, and bootstrapping a fork of emscripten, but that's my plan :-)

**Comment by sbc100** [comment]
It seems reasonable to add support for `posix_spawn` under node, perhaps as past of NODERAWFS.   However, if make is going to spawning all these native processes, why not just have make it self be native too?  I.e. what is the point of have make itself be in wasm/js?

Regarding LLVM you should not need to build it yourself.  You can get it from emsdk.  For information on how to use the emsdk versions of binaryen and llvm with fork of emscripten see https://emscripten.org/docs/tools_reference/emsdk.html#how-do-i-use-my-own-emscripten-fork-with-the-sdk

**Comment by Antony74** [comment]
> what is the point of have make itself be in wasm/js?

I think I answered that [in the reddit thread](https://www.reddit.com/r/node/comments/1fmac0o/comment/lohy5lr/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button), but basically just for the sake of having a clean npm toolchain with no additional dev tools required.

> Regarding LLVM you should not need to build it yourself. You can get it from emsdk.

Thanks!  I didn't get a chance to start looking at that until this evening... I think I'm gradually making headway with it...

**Comment by Antony74** [comment]
Sorry, I'm missing something - the part of the process where emscripten's system/lib/libc (where posix_spawn lives) gets built.

On consecutive docs pages we:

* [Setup the toolchain](https://emscripten.org/docs/building_from_source/toolchain_what_is_needed.html)
* [Configure Emscripten settings](https://emscripten.org/docs/building_from_source/configuring_emscripten_settings.html#)
* [Verify the Emscripten dev environment](https://emscripten.org/docs/building_from_source/verify_emscripten_environment.html)

I keep looking at these pages thinking I've missed a step.

There are no obvious configure or make files except the Makefile and bootstrap.py in the root folder, which don't build libc.  I've introduced an intentional compile error in emscripten_libc_stubs.c and rerun them just to be sure.

I think I'm going to feel terribly silly when I hear the answer to this question, but how do I build libc (and anything that needs building along with it), please?

**Comment by kripken** [comment]
@Antony74 Normally the toolchain builds libc etc. for you automatically (and the emsdk builds contain precompiled libraries) but you can also do it manually using the `embuilder` tool. Do `./embuilder --help` for docs, but basically `embuilder build libc` will build libc, etc.

**Comment by kripken** [comment]
@Antony74 Good idea in https://github.com/Antony74/make.wasm/commit/8da276e74c06dac9e0c3386a5bb098c88129c367 - I think that is worth adding! The only change I'd suggest is to mention `embuilder` there also.

**Comment by Antony74** [comment]
Thanks @kripken, [PR for docs page enhancement](https://github.com/emscripten-core/emscripten/pull/22842) opened.

**Comment by Antony74** [comment]
Above we've been assuming GNU Make uses `posix_spawn` to run the commands within a Makefile.  By default this is not true - it actually uses `vfork` and `execve`.  I think it would be a little bit crazy to try and simulate those in nodejs, so a better approach might be

    #define USE_POSIX_SPAWN

(And at some point I need to figure out why nothing happens when I add `-D USE_POSIX_SPAWN` to my `emcc` command line)

When I do this I get a startup-time error.  i.e. after compile-time but before run-time

        failed to asynchronously prepare wasm: LinkError: WebAssembly.instantiate(): Import #16 module="env" function="__syscall_faccessat" error: function import requires a callable
        Aborted(LinkError: WebAssembly.instantiate(): Import #16 module="env" function="__syscall_faccessat" error: function import requires a callable)
        /git/make/make.js:219
            var e = new WebAssembly.RuntimeError(what);
                    ^
        
        RuntimeError: Aborted(LinkError: WebAssembly.instantiate(): Import #16 module="env" function="__syscall_faccessat" error: function import requires a callable). Build with -sASSERTIONS for more info.
            at abort (/git/make/make.js:219:13)
            at /git/make/make.js:256:13
        
        Node.js v20.18.0

And I'd like to file a bug report saying this is the wrong error at the wrong time, but I'm having trouble isolating it.

GNU Make calls `eaccess` (lib/findprog-in.c).  I can comment them all out and it runs (past that point at least).  But I had to comment quite a lot out to make that happen, so I uncomment just the first one, and the problem returns.  So that's sort of isolated, but when I try to write a test program it compiles, starts, and runs no problem

    // emcc -O2 -s NODERAWFS=1 access.c

    #define _GNU_SOURCE

    #include <stdio.h>
    #include <unistd.h>
    #include <fcntl.h>

    int main() {
        printf("Calling an access function\n");
    //    int result = access("./result.json", X_OK); // Works OK
    //    int result = faccessat(AT_FDCWD, "./result.json", X_OK, AT_EACCESS); // Works OK
        int result = eaccess("./result.json", X_OK); // Also works OK :-(
        printf("Called with result %d\n", result);
        return 0;
    }

So I'm afraid I'm a bit stuck and not sure what to try next.  The commands I'm using to build GNU Make are

    emconfigure ./configure
    emmake make
    find . -name "*.o" -type f | xargs emcc -O2 -s NODERAWFS=1 -o make.js

**Comment by Antony74** [comment]
I can confirm that the equivelent gcc commands produce an apparently well-behaved ./make executable.

    CFLAGS="-DUSE_POSIX_SPAWN" ./configure
    ./make

Apologies for not double-checking this the moment I hit the present error, but I've done it now: this an emscripten-problem(or a my-ignorance-of-emscripten-problem).

**Comment by sbc100** [comment]
> Which also means I've somehow failed to follow the instruction to "Build with -sASSERTIONS for more info" correctly.

`-sASSERTIONS` is a link time flag not a compile time flag.  Are you sure its getting passed to the linker?   Can you see the final link command in the output of `make`?

**Comment by Antony74** [comment]
I'm afraid not.  This issue was not opened by me, and is not solved.  The fact you are asking is an indication I have structured things poorly by not opening additional issues along the way.

#### Problem 1 - can we add posix_spawn to Emscripten so that tools like `make` and `uname` can be built and run with it?
Status: Open.  Depends upon problem 2, and how well the idea fits with the Emscripten project more broadly.  I plan to turn this idea into a concrete proposal, but anticipate the answer will probably be 'no'.

#### Problem 2 - can I build GNU Make in a fork of Emscripten to which posix_spawn has been added?
Status: Open.  Depends on problem 4.  I'm hoping to be able to do this myself, once the matter is ready for me to investigate.

#### Problem 3 - can we similarly build `uname` in a fork of Emscripten?
Status: Open.  Included for completeness only.  My interest is purely in GNU Make so I have no intention of tackling this.  Indeed, I would have to look up what `uname` actually does.

#### Problem 4 - address __syscall_faccessat error
Status: Open.  Depended upon problem 5.  I believe you've just given me enough information about -sASSERTION that I can take another look at this.  Thanks you very much.

#### Problem 5 - how do I build Emscripten's C/C++ system libraries?
Status: Solved.  kripken kindly explained how to do this.  I added some documentation in a PR which was merged after you both approved it.

#### Problem 6 - spellings
Status: Solved.  I spotted a few spelling errors in some README.md files and .c comment banners, which I opened a PR for, and was merged after you both approved it.

### Conclusion

I hope the summary I have given above of this ticket so far has been helpful.  It represents a fairly linear progression which could continue live in this ticket.  However, if I narrow a problem down sufficiently to open another issue for something, then I will, and if you direct me to open another issue then I definitely will!

**Comment by sbc100** [comment]
> #### Problem 2 - can I build GNU Make in a fork of Emscripten to which posix_spawn has been added?
> Status: Open. Depends on problem 4. I'm hoping to be able to do this myself, once the matter is ready for me to investigate.

My hope is that you wouldn't need to fork emscripten, just a `--js-library` that contains `posix_spawn`.   

> #### Problem 3 - can we similarly build `uname` in a fork of Emscripten?
> Status: Open. Included for completeness only. My interest is purely in GNU Make so I have no intention of tackling this. Indeed, I would have to look up what `uname` actually does.

I'm not sure what you mean here.  Emscripten does support uname.   The implementation is here: https://github.com/emscripten-core/emscripten/blob/ffeb76137e765a0023de2635420a4827ca15106d/system/lib/libc/emscripten_syscall_stubs.c#L49-L69

> 
> #### Problem 4 - address __syscall_faccessat error
> Status: Open. Depended upon problem 5. I believe you've just given me enough information about -sASSERTION that I can take another look at this. Thanks you very much.

Emscripten does support this syscall.  The implementation is here: https://github.com/emscripten-core/emscripten/blob/ffeb76137e765a0023de2635420a4827ca15106d/src/library_syscall.js#L941-L967
